### ISSUE TITLE
ArmeriaHttpUtil.toArmeria allocates tons of garbage

### ISSUE THREAD

**Issue by anuraaga** [line/armeria#1112]
Currently, armeria converts netty headers to armeria headers. In some profiling, I noticed tons of `byte[]` garbage is created by `AsciiString.toString()`, because the value type of our headers object is `String`. Would it make sense to change that to `CharSequence` since many use cases are satisfied by it? Alternatively, I'm thinking of putting a cache of `AsciiString` to `String` there to reduce garbage. Out of around 20GB of garbage in a profiling run, 4GB is these headers I think, which is huge for such boring strings like `http`, `https`.

JFR:
[flight_recording_downstream.zip](https://github.com/line/armeria/files/1863483/flight_recording_downstream.zip)

**Comment by anuraaga** [comment]
Sent #1114 as a strawman's proposal - changing the interface of `HttpHeaders` seems pretty intense and not clearly a win.

I'm also wondering whether we should have `DefaultHttpHeaders` contain an `Http2Headers` field instead of creating a new `HttpHeaders` - I noticed a fair bit of allocation of new `HeaderEntry` objects when we copy from the netty object to armeria one. I think it'd be overall better for memory efficiency to wrap instead of copy.

**Comment by trustin** [comment]
Solved partially by #1114, but needs a fundamental fix. Need to think a little bit..

**Comment by trustin** [comment]
gRPC seems to use an alternative `Http2Headers` implementation which serves their use case better: `io.grpc.netty.GrpcHttp2HeadersUtils`, `io.grpc.netty.GrpcHttp2OutboundHeaders`. I think we should follow the same path.

I tried to modify `DefaultHttpHeaders` wrap `DefaultHttp2Headers` to reduce some garbage and was able to gain 2-3% throughput. A better implementation would yield even better numbers.

**Comment by anuraaga** [comment]
Ah I wonder if our branches are similar, I've been slowly working on wrapping too

https://github.com/line/armeria/compare/master...anuraaga:netty-http-headers?expand=1

I wondered if it's worth it without us changing the definition of `HttpHeaders` so the value is `CharSequence` to reduce string conversion and took a break before trying such a large change ;)

I also happened to find https://github.com/netty/netty/blob/4.1/codec-http2/src/main/java/io/netty/handler/codec/http2/ReadOnlyHttp2Headers.java but didn't have good results reproducing the generality of our header conversion logic using it yet.

**Comment by trustin** [comment]
Here's mine: https://github.com/trustin/armeria/commit/06740e7fae4f5fd4885aae907433b78f7d9e44ef

I think we can change the definition of `HttpHeaders` as long as we do not break commonly used methods. Methods like `getByte()` and `getChar()` are of no use in my opinion. ;-)

**Comment by anuraaga** [comment]
Yeah those methods are not useful, but also not a big deal, I was more worried about `get` that always has to call `toString()`, often on an `AsciiString`. But changing the value to `CharSequence` would be highly breaking.
